### DISCUSSION TITLE
PTP with QNEthernet?

### DISCUSSION THREAD

**Discussion by Pragmataraxia** [ssilverman/QNEthernet#34] [Q&A]
Internet articles abound with the news that the Teensy 4.1 "supports PTP".  I have yet to discover any actual implementation of IEEE 1588 for the Teensy 4.1, so I thought I might take a swing at it.  If I can crack it this weekend, my company won't over-engineer the bejesus out of a problem we're trying to solve.

QNEthernet actually seems to have functional UDP multicasting (unlike other interfaces I've tried), and the readme actually mentions PTP frames, so I thought I might start here.  There are a few questions that come to mind though:

1) Am I simply too green to realize that PTP on the Teensy 4.1 is already "solved"?
2) Is QNEthernet suitable for this purpose?  Since it's not asynchronous, a PTP application will need rely on the hardware for timestamping incoming packets... but I haven't the foggiest idea how to access that data.

I don't know who (besides Shawn) will see this discussion, but any tips would be grand.

**Comment by ssilverman** [comment]
I have an API implementation for accessing the IEEE1588 features such as clock adjustment and raw Ethernet frame timestamps, but I haven't released it. If you're interested in talking about it off-list, send me an email.

**Reply by achalkumar076** [reply]
hi shawn, I would like to discuss about PTP on teensy 4.1, what's your email id?

**Reply by ssilverman** [reply]
It’s in the source code headers.
